File: typescript/src/environments.ts
import type { HttpClient } from "./http.js";
import { Environment } from "./environment.js";
import type {
  CreateEnvironmentRequest,
  EnvironmentData,
  ExecRequest,
  ExecResult,
  FileEntry,
  EnvironmentPort,
  CreatePortRequest,
  ListEnvironmentsParams,
} from "./types.js";

export class Environments {
  constructor(private http: HttpClient) {}

  async create(body: CreateEnvironmentRequest): Promise<Environment> {
    const data = await this.http.post<EnvironmentData>(
      this.http.orgPath("/environments"),
      body,
    );
    return new Environment(data, this.http);
  }

  async list(params?: ListEnvironmentsParams): Promise<Environment[]> {
    const qs = new URLSearchParams();
    if (params?.type) qs.set("type", params.type);
    if (params?.state) qs.set("state", params.state);
    const q = qs.toString();
    const path = this.http.orgPath(`/environments${q ? `?${q}` : ""}`);
    const list = await this.http.get<EnvironmentData[]>(path);
    return list.map((d) => new Environment(d, this.http));
  }

  async get(envId: string): Promise<Environment> {
    const data = await this.http.get<EnvironmentData>(
      this.http.orgPath(`/environments/${envId}`),
    );
    return new Environment(data, this.http);
  }

  delete(envId: string): Promise<void> {
    return this.http.delete(this.http.orgPath(`/environments/${envId}`));
  }

  stop(envId: string): Promise<void> {
    return this.http.post(this.http.orgPath(`/environments/${envId}/stop`));
  }

  start(envId: string): Promise<void> {
    return this.http.post(this.http.orgPath(`/environments/${envId}/start`));
  }

  exec(envId: string, req: ExecRequest): Promise<ExecResult> {
    return this.http.post(
      this.http.orgPath(`/environments/${envId}/exec`),
      req,
    );
  }

  listFiles(envId: string, path?: string): Promise<FileEntry[]> {
    const qs = path ? `?path=${encodeURIComponent(path)}` : "";
    return this.http.get(
      this.http.orgPath(`/environments/${envId}/files${qs}`),
    );
  }

  async uploadFile(
    envId: string,
    src: Uint8Array,
    remotePath: string,
  ): Promise<void> {
    const qs = `?path=${encodeURIComponent(remotePath)}`;
    await this.http.requestRaw(
      "POST",
      this.http.orgPath(`/environments/${envId}/files/upload${qs}`),
      src,
      "application/octet-stream",
    );
  }

  async downloadFile(envId: string, remotePath: string): Promise<ArrayBuffer> {
    const qs = `?path=${encodeURIComponent(remotePath)}`;
    const res = await this.http.requestRaw(
      "GET",
      this.http.orgPath(`/environments/${envId}/files/download${qs}`),
    );
    return res.arrayBuffer();
  }

  listPorts(envId: string): Promise<EnvironmentPort[]> {
    return this.http.get(this.http.orgPath(`/environments/${envId}/ports`));
  }

  createPort(envId: string, req: CreatePortRequest): Promise<EnvironmentPort> {
    return this.http.post(
      this.http.orgPath(`/environments/${envId}/ports`),
      req,
    );
  }

  deletePort(envId: string, portId: string): Promise<void> {
    return this.http.delete(
      this.http.orgPath(`/environments/${envId}/ports/${portId}`),
    );
  }
}

File: go/environments.go
package codewire

import (
	"context"
	"fmt"
	"io"
	"net/url"

	"codewire.sh/sdk-go/internal/api"
)

type EnvironmentService struct {
	client *Client
}

func (s *EnvironmentService) Create(ctx context.Context, req api.CreateEnvironmentBody) (*Environment, error) {
	path, err := s.client.orgPath("/environments")
	if err != nil {
		return nil, err
	}
	var data api.Environment
	if err := s.client.do(ctx, "POST", path, req, &data); err != nil {
		return nil, err
	}
	return s.wrap(data), nil
}

func (s *EnvironmentService) List(ctx context.Context, params *api.ListEnvironmentsParams) ([]*Environment, error) {
	path, err := s.client.orgPath("/environments")
	if err != nil {
		return nil, err
	}
	if params != nil {
		q := url.Values{}
		if params.Type != nil {
			q.Set("type", *params.Type)
		}
		if params.State != nil {
			q.Set("state", *params.State)
		}
		if qs := q.Encode(); qs != "" {
			path += "?" + qs
		}
	}
	var list []api.Environment
	if err := s.client.do(ctx, "GET", path, nil, &list); err != nil {
		return nil, err
	}
	envs := make([]*Environment, len(list))
	for i, d := range list {
		envs[i] = s.wrap(d)
	}
	return envs, nil
}

func (s *EnvironmentService) Get(ctx context.Context, envID string) (*Environment, error) {
	path, err := s.client.orgPath("/environments/%s", envID)
	if err != nil {
		return nil, err
	}
	var data api.Environment
	if err := s.client.do(ctx, "GET", path, nil, &data); err != nil {
		return nil, err
	}
	return s.wrap(data), nil
}

func (s *EnvironmentService) Delete(ctx context.Context, envID string) error {
	path, err := s.client.orgPath("/environments/%s", envID)
	if err != nil {
		return err
	}
	return s.client.do(ctx, "DELETE", path, nil, nil)
}

func (s *EnvironmentService) Stop(ctx context.Context, envID string) error {
	path, err := s.client.orgPath("/environments/%s/stop", envID)
	if err != nil {
		return err
	}
	return s.client.do(ctx, "POST", path, nil, nil)
}

func (s *EnvironmentService) Start(ctx context.Context, envID string) error {
	path, err := s.client.orgPath("/environments/%s/start", envID)
	if err != nil {
		return err
	}
	return s.client.do(ctx, "POST", path, nil, nil)
}

func (s *EnvironmentService) Exec(ctx context.Context, envID string, req api.ExecBody) (*api.ExecResult, error) {
	path, err := s.client.orgPath("/environments/%s/exec", envID)
	if err != nil {
		return nil, err
	}
	var result api.ExecResult
	if err := s.client.do(ctx, "POST", path, req, &result); err != nil {
		return nil, err
	}
	return &result, nil
}

func (s *EnvironmentService) ListFiles(ctx context.Context, envID string, dirPath *string) ([]api.FileEntry, error) {
	path, err := s.client.orgPath("/environments/%s/files", envID)
	if err != nil {
		return nil, err
	}
	if dirPath != nil {
		path += "?path=" + url.QueryEscape(*dirPath)
	}
	var files []api.FileEntry
	if err := s.client.do(ctx, "GET", path, nil, &files); err != nil {
		return nil, err
	}
	return files, nil
}

func (s *EnvironmentService) UploadFile(ctx context.Context, envID string, src io.Reader, remotePath string) error {
	path, err := s.client.orgPath("/environments/%s/files/upload?path=%s", envID, url.QueryEscape(remotePath))
	if err != nil {
		return err
	}
	resp, err := s.client.doRaw(ctx, "POST", path, src, "application/octet-stream")
	if err != nil {
		return err
	}
	resp.Body.Close()
	return nil
}

func (s *EnvironmentService) DownloadFile(ctx context.Context, envID string, remotePath string) (io.ReadCloser, error) {
	path, err := s.client.orgPath("/environments/%s/files/download?path=%s", envID, url.QueryEscape(remotePath))
	if err != nil {
		return nil, err
	}
	resp, err := s.client.doRaw(ctx, "GET", path, nil, "")
	if err != nil {
		return nil, err
	}
	return resp.Body, nil
}

func (s *EnvironmentService) ListPorts(ctx context.Context, envID string) ([]api.EnvironmentPort, error) {
	path, err := s.client.orgPath("/environments/%s/ports", envID)
	if err != nil {
		return nil, err
	}
	var ports []api.EnvironmentPort
	if err := s.client.do(ctx, "GET", path, nil, &ports); err != nil {
		return nil, err
	}
	return ports, nil
}

func (s *EnvironmentService) CreatePort(ctx context.Context, envID string, req api.CreatePortBody) (*api.EnvironmentPort, error) {
	path, err := s.client.orgPath("/environments/%s/ports", envID)
	if err != nil {
		return nil, err
	}
	var port api.EnvironmentPort
	if err := s.client.do(ctx, "POST", path, req, &port); err != nil {
		return nil, err
	}
	return &port, nil
}

func (s *EnvironmentService) DeletePort(ctx context.Context, envID, portID string) error {
	path, err := s.client.orgPath("/environments/%s/ports/%s", envID, portID)
	if err != nil {
		return err
	}
	return s.client.do(ctx, "DELETE", path, nil, nil)
}

func (s *EnvironmentService) wrap(data api.Environment) *Environment {
	orgID, _ := s.client.resolveOrgID()
	return &Environment{
		Environment: data,
		client:      s.client,
		orgID:       orgID,
	}
}

func orgPath(orgID, pattern string, args ...any) string {
	allArgs := make([]any, 0, len(args)+1)
	allArgs = append(allArgs, orgID)
	allArgs = append(allArgs, args...)
	return fmt.Sprintf("/organizations/%s"+pattern, allArgs...)
}
